### PR TITLE
Phase 12-D: CUDA batched LM-head argmax

### PR DESCRIPTION
**Pull request by ericflo**
## Summary

- Adds `kiln-lm-head-kernel`, a CUDA BF16 batched LM-head greedy argmax path that returns one token id per decode row without materializing `[batch, 1, vocab]` logits.
- Wires CUDA `supports_linear_decode_argmax_batch()` / `linear_decode_argmax_batch()` so PR #1004's safety gate can enter contiguous-batched greedy decode when the new path supports the shape.
- Keeps a kill switch: `KILN_DISABLE_CUDA_LM_HEAD_ARGMAX_BATCH=1`.

## Validation

A6000 pool lease `pod-ca3fe95dc6e032e97c23f3c0`, kiln-runpod image, `KILN_CUDA_ARCHS=86`, W4A16, CUDA graphs, prefix cache, batching engine, `KILN_DECODE_PARALLELISM=8`.

- PASS: `cargo fmt --check`
- PASS: `KILN_CUDA_ARCHS=86 cargo build --release --features cuda --bin kiln`
- PASS: `KILN_CUDA_ARCHS=86 cargo nextest run --release -p kiln-lm-head-kernel test_cuda_lm_head_argmax`
- PASS: `KILN_CUDA_ARCHS=86 cargo nextest run --release -p kiln-model test_cuda_backend_lm_head_argmax_batch_matches_materialized_logits --features cuda`
- PASS: `KILN_CUDA_ARCHS=86 cargo nextest run --release -p kiln-model cuda_graph --features cuda`
- NOTE: `cargo nextest run --release -p kiln-model batched_decode --features cuda` still matches no tests; command was run with the same known no-op behavior as PR #1004.
- PASS reliability smoke: c=8 Shape-A-style streaming smoke 32/32 `finish_reason=length`, 0 errors, 0 LM-head/CUDA OOMs, 0 illegal-address logs.
- PASS parity: focused CUDA kernel + backend argmax parity tests cover batch sizes 1/2/4/8 against materialized logits.

## Full sweep verdict

**ACCEPTANCE: FAIL / do not merge.**

Report: `b2://clouderic/diagnostics/kiln-phase12-d-cuda-argmax-validation-2026-05-09/REPORT.md`

- Shape A c=1: 44.37 tok/s vs floor 49.30 — FAIL
- Shape A c=8: 34.13 tok/s vs baseline recovery gate 62.45 — FAIL
- OOM log matches: 0 — PASS
- Illegal-address log matches: 0 — PASS
- HTTP errors: 0 in counted rows — PASS
- c>=8 non-length finishes: 0 in counted rows — PASS

Diagnosis: the CUDA argmax path fixes the missing backend capability/reliability problem, but the current cuBLAS BF16 GEMM + GPU reduce implementation is slower than PR #1004's rowwise reliable fallback and does not restore throughput. Leaving this draft as evidence; auto-merge is intentionally not enabled.

### CODE REVIEW

**Comment by ericflo** [comment]
Phase 12-D CUDA batched LM-head argmax implementation opened as draft PR #1005: https://github.com/ericflo/kiln/pull/1005

Verdict: **ACCEPTANCE FAIL / do not merge**.

Validation report: `b2://clouderic/diagnostics/kiln-phase12-d-cuda-argmax-validation-2026-05-09/REPORT.md`

Key results at validated SHA `db75f22dad1b7e3d475026c39a3abd161a649d84`:
- Parity: PASS — focused CUDA kernel + backend argmax parity tests pass for batch sizes 1/2/4/8 against materialized logits.
- Reliability: PASS — 0 LM-head/CUDA OOM log matches, 0 illegal-address matches, 0 counted HTTP errors, all counted c>=8 rows finished `length`.
- Shape A c=1: 44.37 tok/s vs 49.30 floor — FAIL.
- Shape A c=8: 34.13 tok/s vs 62.45 baseline recovery gate — FAIL.

Diagnosis: the backend capability is now implemented, but the current cuBLAS BF16 GEMM + GPU reduce argmax implementation is slower than PR #1004's rowwise reliable fallback. Auto-merge was not enabled.

**Comment by ericflo** [comment]
Closing this draft per Phase 12 regression discipline: validation at db75f22 showed reliability/parity pass but throughput failed both gates (Shape A c=1 44.37 < 49.30 floor; c=8 34.13 < 62.45 recovery baseline). The cuBLAS BF16 GEMM + GPU reduce argmax path is slower than the reliable rowwise fallback, so this branch should not be salvaged.
